src/finchlite/finch_assembly/dataflow.py

```diff
@@ -0,0 +1,150 @@
+from abc import abstractmethod
+
+from ..symbolic import DataFlowAnalysis
+from ..util import qual_str
+from .cfg_builder import assembly_build_cfg
+from .nodes import (
+    AssemblyNode,
+    Assert,
+    Assign,
+    Call,
+    Literal,
+    TaggedVariable,
+    Variable,
+)
+
+
+def assembly_copy_propagation(node: AssemblyNode):
+    ctx = AssemblyCopyPropagation(assembly_build_cfg(node))
+    ctx.analyze()
+    return ctx
+
+
+class AbstractAssemblyDataflow(DataFlowAnalysis):
+    def stmt_str(self, stmt, state: dict) -> str:
+        """Annotate a statement with lattice values."""
+        match stmt:
+            case Assign(lhs, rhs):
+                rhs_str = self._expr_str(rhs, state)
+                return f"{str(lhs)} = {rhs_str}"
+            case Assert(exp):
+                exp_str = self._expr_str(exp, state)
+                return f"assert({exp_str})"
+            case _:
+                return str(stmt)
+
+    def _expr_str(self, expr, state: dict) -> str:
+        """Annotate an expression with lattice values."""
+        match expr:
+            case Literal(value):
+                return qual_str(value)
+            case TaggedVariable(Variable(name, _), id):
+                var_str = f"{name}_{id}"
+                if name in state:
+                    return f"{var_str}: {self.print_lattice_value(name, state)}"
+                return var_str
+            case Variable(name, _):
+                if name in state:
+                    return f"{name}: {self.print_lattice_value(name, state)}"
+                return str(name)
+            case Call(Literal(_) as lit, args):
+                annotated_args = [self._expr_str(arg, state) for arg in args]
+                return f"{qual_str(lit.val)}({', '.join(annotated_args)})"
+            case _:
+                return str(expr)
+
+    @abstractmethod
+    def print_lattice_value(self, name, state: dict) -> str:
+        """Format the lattice value associated with a variable for annotation."""
+        ...
+
+
+class AssemblyCopyPropagation(AbstractAssemblyDataflow):
+    def direction(self) -> str:
+        """
+        Copy propagation is a forward analysis.
+        """
+        return "forward"
+
+    def transfer(self, stmts, state: dict) -> dict:
+        new_state = state.copy()
+
+        for stmt in stmts:
+            match stmt:
+                case Assign(lhs, rhs):
```

Comment:
Don't we need to check if the rhs is a variable? If it's a variable, we should check whether that variable has a copy, and use that instead. I'm asking that you do the "searching for root variable" calculation that happened in the print function in the transfer function instead. That should simplify the printing.